File: src/bin/0104-maximum-depth-of-binary-tree.py
# @leetup=info id=104 lang=python3 slug=maximum-depth-of-binary-tree



from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @leetup=code
class Solution(object):
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(node: TreeNode) -> int:
            if not node:
                return 0
            return 1 + max(dfs(node.left), dfs(node.right))

        return dfs(root)


# @leetup=code

File: src/bin/0104-maximum-depth-of-binary-tree.rs
// @leetup=info id=104 lang=rust slug=maximum-depth-of-binary-tree


// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @leetup=code
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
            if let Some(root) = root {
                1 + dfs(root.borrow().left.clone()).max(dfs(root.borrow().right.clone()))
            } else {
                0
            }
        }

        dfs(root)
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::max_depth;
}
